src/q4/os-SINF1252/exam/2019/Juin/Majeure/os-SINF1252-exam-2019-Juin-Majeure.tex

```diff
@@ -0,0 +1,119 @@
+Écrivez l'algorithme de Peterson (ou variante) en pseudo-code.
+\begin{solution}
+\lstinputlisting{../../../../summary/code/peterson.c}%autant ne pas dupliquer les codes
+\end{solution}
+
+\section{Question 3 [2 points]}
+Pour un processus, quelles sont les pages qui ont les permissions RW- et quels sont celles dans les permissions sont R-X?
+\begin{solution}
+\textbf{---TO DO---}
+\end{solution}
+
+
+\section{Question 4 [4 points]}
+Dans quelle zone de la mémoire se trouvent les variables a, b, c, r, g, k, le poiteur ptr (ainsi que la zone vers laquelle il pointe) et les codes des fonctions f, main, printf? Donnez aussi l'état de la stack lors de l'exécution de la fonction f, à l'endroit du macro ---ICI---. 
+\lstinputlisting{Code/zoneMemoire.c}
+
+\begin{solution}
+\textbf{---TO COMPLETE---}
+
+a, b, r, ptr, g, k sont sur la stack
+
+Le contenu de ptr est sur le heap.
+
+c est avec les variables globales.
+
+La fonction f et la main sont dans le segment texte
+
+La fonction standard printf est dans les données dynamiques, entre la stack et le heap.
+
+Afin de représenter la stack, il faut se rappeler que la stack dédiée à la fonction f conserve celle de la fonction main. En fait, à l'appel de la fonction f, on pousse sur la stack les arguments de f (en commençant par le dernier) ainsi que son adresse de retour. Seulement ensuite, viennent les valeurs telles que \og r \fg{}, provenant de l'exécution de f.
+\end{solution}
+
+\section{Question 5 [2 points]}
+Expliquez les fonctionnement de la fonction pthread\_create notamment en expliquant ce qu'il se passe dans la table des pages.
+
+\begin{solution}
+\textbf{---TO DO---}
+\end{solution}
+
+
+\part{Partie pratique}
+\section{Question 1 [10\% de la partie inginious]}
+Traduction d'une fonction en assembleur
+\begin{lstlisting}
+f : 	movl 4(%esp), %eax
+	movl 8(%esp), %ebx
+	movl (%ebx), %ecx
+	addl %ecx ,%eax
+	movl 12(%esp), %ecx
+	addl %ecx, eax
+	ret
+\end{lstlisting}
+Ce code n'est pas le code original mais s'en rapproche, si un de vous a passé l'examen et s'en rappelle... N'hésitez pas à l'éditer!
+\begin{solution}
+Le point important de cette exercice est l'utilisation d'un pointeur pour le deuxième paramètre! En effet, la présence de parenthèse indique le mode d'adressage indirect.
+\end{solution}
+
+
+\section{Question 2 [45\% de la partie inginious]}
+Ecrire la fonction « add » d’une liste simplement chaînée.
+
+\begin{solution}
+\textbf{---TO DO---}
+\end{solution}
+
+\section{Question 3 [45\% de la partie inginious]}
+Ecrire une fonction qui réalise la même chose que la commande « cp » d’un shell.
+bash (réalise une copie d’un fichier) et qui conserve les mêmes permissions sur les 2 fichiers, en utilisant uniquement les appels systèmes open, read, write, stat et close. Il est interdit d'utiliser un buffer de taille supérieure à 100 bytes ainsi que de faire appel à la fonction \og system \fg{}.
+
+\begin{solution}
+\textbf{---TO DO---}
+\end{solution}
+
+\paragraph{Note}
```

Comment:
Cette note est pas problématique à priori, mais en général, si quelqu'un a contribué de façon significative au document, c'est normal d'inclure son nom dans les auteurs (avec `\and` entre).